src/librustc_typeck/check/demand.rs

```diff
@@ -58,7 +59,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let parent_node = self.tcx.map.get(self.tcx.map.get_parent_node(expr.id));
+            let implicit_return = match parent_node {
+                Node::NodeBlock(_) => true,
+                _ => false,
+            };
+            self.report_mismatched_types_for_return(origin, expected, expr_ty, e, implicit_return);
```

Comment:
Hmm, thinking about it more, could the logic be moved here? Or even better, in `check_block_with_expectation`?
See #37412 for how I moved a note there. Maybe I should get that PR merged first, would make this easier.